synapse/cryotank.py

```diff
@@ -571,3 +589,644 @@ def init(self, name, conf=None, timeout=None):
+
+            record_tuples = self.cryotank.rows(self._next_offset, self._chunk_sz)
+            norm_gen = self._normalize_records(record_tuples)
+            rowcount = self._writeIndices(norm_gen)
+
+            self._removeSome()
+            # logger.debug('Processed %d rows', rowcount)
+            if not rowcount and not self._meta.deleting:
+                if stillworktodo is True:
+                    # logger.info('Completely caught up with indexing')
+                    stillworktodo = False
+            else:
+                stillworktodo = True
+            # loop_end_t = loop_start_t + self.MAX_WAIT_S
+
+    @_inWorker
+    def addIndex(self, prop, syntype, datapath, *args):
+        '''
+        Adds an index to the cryotank.
+
+        Args:
+            prop (str):  the name of the property this will be stored as in the normalized record
+            syntype (str):  the synapse type this will be interpreted as
+            datapath (str):  the datapath spec against which the raw record is run to extract a single field
+        that is passed to the type normalizer.
+            *args (str):  additional datapaths that will be tried in order if the first isn't present.
+        Returns:
+            None
+        Note:
+            Additional datapaths will be tried iff prior datapaths are not present, and *not* if
+            the normalization fails.
+        '''
+        return self._meta.addIndex(prop, syntype, datapath, args)
+
+    @_inWorker
+    def delIndex(self, prop):
+        '''
+        Deletes an index
+
+        Args:
+            prop (str): the (normalized) property name
+        Returns:
+            None
+        '''
+        return self._meta.delIndex(prop)
+
+    @_inWorker
+    def pauseIndex(self, prop=None):
+        '''
+        Temporarily stop indexing one or all indices.
+
+        Args:
+            prop: (Optional[str]):  the index to stop indexing, or if None, indicate to stop all indices
+        Returns:
+            None
+        Note:
+            Pausing is not persistent.  Restarting the process will resume indexing.
+        '''
+        return self._meta.pauseIndex(prop)
+
+    @_inWorker
+    def resumeIndex(self, prop=None):
+        '''
+        Undo a pauseIndex.
+        Args:
+            prop: (Optional[str]):  the index to start indexing, or if None, indicate to resume all indices
+        Returns:
+            None
+        '''
+        return self._meta.resumeIndex(prop)
+
+    @_inWorker
+    def getIndices(self):
+        '''
+        Args:
+            None
+        Returns
+            List[Dict[str: Any]]: all the indices with progress and statistics
+        '''
+        idxs = {iid: dict(metaentry.asdict()) for iid, metaentry in self._meta.indices.items()}
+        for iid in idxs:
+            idxs[iid].update(self._meta.progresses.get(iid, {}))
+        return list(idxs.values())
+
+    def _iterrows(self, prop, valu, exact=False):
```

Comment:
Should callers form an xact and hand in here?  Are there cases where we'll want to call _iterrows multiple times in quick succession?  Been trying to spot places where the internal APIs could be used to allow more granular xact boundary control.

Reply:
For a user of the API to form a transaction, he'd need access to the database object or we'd need to make a helper for him.